### ISSUE TITLE
KonstruktFileActionResult downloads corrupted Word DOCX

### ISSUE THREAD

**Issue by matt-tolliday-iss** [outfielddigital/konstrukt#58]
**Describe the bug**
I'm unsure if this is unrelated to KonstruktFileActionResult but, using the following code, I am reading a Media file from the Media Library, via a KonstruktAction, using a KonstruktPicker on a "Media Picker" type in the ActionSettings, which then loads the Media file into memory, and is supposed to download it to the user executing the bulk action.

<img width="997" alt="image" src="https://github.com/outfielddigital/konstrukt/assets/80101704/20a21131-1215-4306-8345-2e9d4adb1b18">

If i use the commented code to write to the file system directly, then my file is uncorupted and as expected. But using the FileActionResult mechanism produces the following issues when opening:

<img width="694" alt="image" src="https://github.com/outfielddigital/konstrukt/assets/80101704/e063f7b5-16f0-4851-b6b1-6e7524c4e822">

Selecting yes

<img width="304" alt="image" src="https://github.com/outfielddigital/konstrukt/assets/80101704/2411f1ac-081e-4763-a948-f459e0131505">

The contentType it infers is: "application/vnd.openxmlformats-officedocument.wordprocessingml.document"

The resulting file sizes are drastically different on the file system too.

Does the File Action Result set the Content-Length header? I've read that can cause issues.

**Environment (please complete the following information):**
 - Umbraco Version [10.3.2]
 - Product Version [1.6.4]

**Comment by mattbrailsford** [comment]
Hey matt. 

I guess the thing that throws me is the "wildly different file sizes". My thoughts really are whether the document is encoded differently to what the `KonstruktFileActionResult` is returning, but if that was the case, I wouldn't really expect them to be massively different files sizes.

Under the hood all that really happens is the stream is passed to a `File` method on the controller to return a file so we aren't really doing anything with the stream ourselves.

    return File(fileResult.FileStream, fileResult.FileType, fileResult.FileName);

The only other option I could think to try is the `KonstruktFileActionResult` has a constructor that can take the file bytes, rather than a stream, so you could try loading the entire file into memory and pass the bytes through, but again, we don't really do any processing on that, it also just gets passed to the `File` method

    return File(fileResult.FileBytes, fileResult.FileType, fileResult.FileName);

**Comment by AanEasyflow** [comment]
Hi @mattbrailsford

I actually have the same problem. If I create a xml file, and save the file directly from the code, I can open it without problems. If I send the file as either byte array or just as stream, the file is bigger than the one I saved on disk. I can see in the response that it returns with the right filesize, but the file saved on disk is alot bigger.

**Comment by mattbrailsford** [comment]
Interesting. Do you have any links to what you were reading up on?

**Comment by AanEasyflow** [comment]
There are many different solutions out their, but none of which I have got them working. Maybe make another class where you can pass an endpoint that you call in the frontend after accepting the popup? Then you wouldn't have to handle the file request yourself. Don't know if it is a bad solution, but it is one :)

**Comment by mattbrailsford** [comment]
Hmm, I'll have to do some investigating post Codegarden

**Comment by mattbrailsford** [comment]
Hey @AanEasyflow 

Unfortunately not yet. We only got back in the office yesterday after CG and as you can imagine there is lot on with the launch of Umbraco Commerce. I'll try and get to it as soon as I can.

**Comment by AanEasyflow** [comment]
Alright! Thanks alot @mattbrailsford 
Really appreciated :)

**Comment by AanEasyflow** [comment]
Hi @mattbrailsford 

I was just adding some new repositories, when I made a new discovery. The file served, does work depending on where the action result is called from.

I have tested this out, only using the same static file from disk. And depending on where i serve the file from, I can open it.

When calling the "KonstruktAction" using the:
1) KonstruktActionType.ContainerMenu 
2) KonstruktActionType.EntityMenu
It cannot open the file and the file size is wrong when served to me.


When calling the "KonstruktAction" using the:
3) KonstruktActionType.Bulk 
4) KonstruktActionType.Row
I CAN open the file and the file size is right when served to me.

No matter where I make this call. I is the same result I return with a static file from disk:
 "new KonstruktFileActionResult(true, byteInfo, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "test.xlsx")"

Hope this might help you :)


See screenshot with explanations:
![bug-file-request-konstrukt](https://github.com/outfielddigital/konstrukt/assets/122968130/673d6e61-64f8-46cf-92f4-7be2bc7c65c5)

**Comment by matt-tolliday-iss** [comment]
Just to follow up on this from my side. I am using documents as retrieved from the Azure Blob plugin - and we are using the BULK menu item type for the Action, where the file is not downloading correctly. Which seems slightly in-contra to the above.

**Comment by mattbrailsford** [comment]
Doh! 🤦‍♂️

**Comment by AanEasyflow** [comment]
Hi @matt-tolliday-iss 

Yeah that sounds contradictive. Have you tried to make a normal API controller where you serve the file using: 
_return File(result.FileStream, result.FileType, result.FileName)_
Using the same code you linked above and does this work? It would be a good test if you tried to see, if you could serve the file with a normal API request. The reason I looked into it, was because I was able to open the file when I downloaded the file, using a normal API request.

**Comment by AanEasyflow** [comment]
Hi @mattbrailsford 
Have you had the time to look at the problem?

Regards Andreas

**Comment by mattbrailsford** [comment]
Not yet, I was hoping @matt-tolliday-iss might attempt your suggestion and see if he saw the same. 

I should have some time next week to review

**Comment by mattbrailsford** [comment]
So I've been able to replicate this and I think it does come down to how we handle the file response in javascript to trigger the download (that and there was another bug where a filetype wasn't being passed through for menu actions correctly and so the right response type wasn't being set). I've updated both of these and in my tests it does now download a file correctly.

I've pushed an update to our unstable feed at https://nuget.outfield.digital/unstable/v3/index.json so there should be a new 1.6.5 beta build shortly. If anyone wants to test this and see if it resolves the issue for you, that would be appreciated.
